sdk/scheduler/src/main/java/org/apache/mesos/offer/constrain/MaxPerAttributeGenerator.java

```diff
@@ -124,4 +95,84 @@ public boolean equals(Object o) {
     public int hashCode() {
         return HashCodeBuilder.reflectionHashCode(this);
     }
+
+    /**
+     * Implementation of max attribute counts. Considers attribute usage of tasks in the cluster to
+     * determine whether the provided task can be launched against a given offer. This rule is
+     * checked at offer evaluation time in order to check that the task isn't being double-counted
+     * against itself in a config update scenario.
+     */
+    private static class MaxPerAttributeRule implements PlacementRule {
+
+        private final int maxTasksPerSelectedAttribute;
+        private final AttributeSelector attributeSelector;
+        private final Collection<TaskInfo> tasks;
+
+        private MaxPerAttributeRule(
+                int maxTasksPerSelectedAttribute,
+                AttributeSelector attributeSelector,
+                Collection<TaskInfo> tasks) {
+            this.maxTasksPerSelectedAttribute = maxTasksPerSelectedAttribute;
+            this.attributeSelector = attributeSelector;
+            this.tasks = tasks;
```

Comment:
I'm a little confused by Tasks are being passed up to a PlacementRule.  Tasks are available at the PlacementRuleGenerator level.  Couldn't this implementation reduce to a PlacementRuleGenerator that generates an ORed list of AgentRules?